### DISCUSSION TITLE
Update cmap when changing scalars

### DISCUSSION THREAD

**Discussion by teliason** [pyvista/pyvista#4131] [Q&A] [answered]
I have an application where I have a mesh with two different sets of scalar data that are mapped to a single mesh.  I want the user to be able to toggle between them to visualize this scalar data, but these two datasets are completely different ranges and therefore need different color maps for proper visualization.  I can toggle scalars with no issue, but have not been able to figure out how to update the cmap that the plotter is using without removing the mesh and reimporting it with a new cmap.  

I'd prefer to not have to do this as my application allows the user to interact with the mesh (hiding elements,) prior to turning the scalar values, which would be undone by removing and reimporting the mesh.  Any guidance you can give on how to approach updating the cmap in place would be much appreciated. 

Simple example code below that shows toggling between two random scalar datasets.  

```
import pyvista as pv
from numpy import random


scalar = True

def change_scalar():
    global scalar
    
    if scalar:
        pl.mesh.set_active_scalars('values2')
        scalar = False
        pl.update()
    else:
        pl.mesh.set_active_scalars('values')
        scalar = True
        pl.update()
       

mesh = pv.Sphere()   

values = random.rand(mesh.number_of_cells)
values2 = random.rand(mesh.number_of_cells)

 
mesh.cell_data['values'] = values
mesh.cell_data['values2'] = values2
 

pl = pv.Plotter()
pl.add_mesh(mesh, scalars= 'values', show_edges=True)


pl.add_key_event('t', change_scalar)

pl.show()
```

**Comment by banesullivan** [comment, marked as answer]
After plotting, you'll simply modify the parameters of the mapper:

```py
...
actor = pl.add_mesh(mesh, scalars= 'values', show_edges=True)
...
# Change scalars and range
actor.mapper.array_name = "array name"
actor.mapper.scalar_range = mesh.get_data_range("array name")
```

To demonstrate this, I'm adding an example in https://github.com/pyvista/pyvista/pull/4138

**Reply by teliason** [reply]
Thank you that is helpful to be able to get access to the scalar range and update it dynamically.  Is the cmap definition also available in a similar manner?

In addition to changing the range I would also like to use a different color scheme between the two different scalar values.

**Reply by teliason** [reply]
I was able to figure out how to do what I needed with actor.mapper.set_scalars().

Thanks for pointing me in the right direction with the mapper.

**Reply by banesullivan** [reply]
You can simply change the colormap via the mapper's lookup table:

```py
import pyvista as pv

mesh = pv.Wavelet()

pl = pv.Plotter()
actor = pl.add_mesh(mesh, show_edges=True)
pl.show()

...
# Change colormap
actor.mapper.lookup_table.cmap = 'coolwarm'
```

**Reply by hakonanes** [reply]
Regarding your example just above here, @banesullivan: The mesh colors are not updated with the call to change the colormap. You use `server = trame.app.get_server(); server.controller.view_update()` in your example addressing this use case in #4138. Can I update the view without using trame? `Plotter.update()` does not work, as in

```python
import pyvista as pv

mesh = pv.Wavelet()

pl = pv.Plotter()
actor = pl.add_mesh(mesh, show_edges=True)
pl.show(jupyter_backend="static")

# Change colormap
actor.mapper.lookup_table.cmap = 'coolwarm'

# Try to update view
pl.update()
```

**Reply by banesullivan** [reply]
FYI, I re-ran into this issue not too long ago: https://github.com/pyvista/pyvista/pull/5998

First, `jupyter_backend="static"` will not be able to be updated.. You must use a trame-based backend for interactive updates.

In a Jupyter notebook, the following works for me:


```py
import pyvista as pv
pv.set_jupyter_backend('server')

mesh = pv.Wavelet()

pl = pv.Plotter()
actor = pl.add_mesh(mesh, show_edges=True)
pl.show()
```

In a separate cell, later:

```py
...
# Change colormap
actor.mapper.lookup_table.cmap = 'coolwarm'
pl.update()  # The trame backend should see this update and reflect the changes
```

**Comment by huyidao625** [comment]
Banesullivan's answer is correct.

Don't forget to change the` active_scalars_name` of the `mapper dataset`  after modifying the scalars and ranges.

`actor.mapper.array_name = "array name"`
`actor.mapper.scalar_range = mesh.get_data_range("array name")`
`actor.mapper.dataset.active_scalars_name = "array name"`
